src/compiler/filesystem.cc

```diff
@@ -141,13 +141,14 @@ void Filesystem::canonicalize(char* path) {
       path[canonical_pos++] = path[i++];
     }
   }
-  // Drop trailing path seperator.
-  // There can only be one.
-  if (path[canonical_pos - 1] == path_separator()) {
+  // Drop trailing path seperator unless it's the root.
+  path[canonical_pos] = '\0';
+  if (!(is_absolute && is_root(path)) && path[canonical_pos - 1] == path_separator()) {
+    // There can only be one.
     canonical_pos--;
   }
   if (canonical_pos == 0) {
```

Comment:
Can this happen after stripping the last separator? That feels weird.

Reply:
Moved the `canonical_pos == 0` check before the 'is_root` check and made the `is_root` check an 'else'.
This also avoids accessing the path at `-1` which could otherwise happen.
done.